core/src/test/java/io/grpc/internal/ServerImplTest.java

```diff
@@ -822,9 +825,55 @@ public void cancelled(Context context) {
     assertNotNull(streamListener);
 
     streamListener.onReady();
+    assertEquals(1, executor.runDueTasks());
+    return streamListener;
+  }
+
+  @Test
+  public void testClientClose_cancelTriggersImmediateCancellation() throws Exception {
+    final AtomicBoolean contextCancelled = new AtomicBoolean(false);
+    final AtomicReference<Context> context = new AtomicReference<Context>();
+    final AtomicReference<ServerCall<String, Integer>> callReference
+        = new AtomicReference<ServerCall<String, Integer>>();
+
+    ServerStreamListener streamListener = testClientClose_setup(callReference,
+        context, contextCancelled);
+
+    // For close status being non OK:
+    // isCancelled is expected to be true immediately after calling closed(), without needing
+    // to wait for the main executor to run any tasks.
+    assertFalse(callReference.get().isCancelled());
+    assertFalse(context.get().isCancelled());
     streamListener.closed(Status.CANCELLED);
+    assertEquals(1, cancelExecutor.runDueTasks());
```

Comment:
IIUC, this is the reason you added `cancelExecutor`, because you only want the cancellation runnable to run, and leave out the stream listeners. Instead of having for-test mechanism in the implementation, you may consider this alternative approach:

Give the cancellation runnable a class, and add `runDueTasks(TaskFilter)` to `FakeClock` that runs selected runnables:
```java
interface TaskFilter {
  boolean shouldRun(Runnable task);
}
```